### ISSUE TITLE
Systematically determine `min_dist` and `n_neighbors`

### ISSUE THREAD

**Issue by NTNguyen13** [lmcinnes/umap#182]
I tested multiple combination of `min_dist` and `n_neighbor` on my data, I found that a suitable combination of those hyperparameters can separate the all the cluster(according to labels). I wonder how to determine those hyperparameters systematically, not by visualization of the data after embedding.

**Comment by lmcinnes** [comment]
If you have specific labels to measure against you could cluster the embedding (with, say, hdbscan) and look at the adjusted Rand score (``adjusted_rand_score`` in ``sklearn.metrics``). of the clustering against the labels. Maximising that score would be giving an embedding that most closely matches your labels.

Personally I would be a little wary of such hyper-parameter searching, but a lot depends on your experiment and the objectives you have in mind. Best of luck.

**Comment by NTNguyen13** [comment]
yes, I'm actually doing the Hyperparameter Optimization with ARI metric for the particular problem. But I would like to know if there is a more general approach, such as in problems without the labels.

**Comment by lmcinnes** [comment]
If there are no labels then there is not "truth", so no, there isn't a general method that works without labels.

**Comment by MattWenham** [comment]
This has got me thinking, and that can be a dangerous or wonderful thing depending on circumstances.

I could envisage a scheme which had `min_dist`, `n_neighbor` and `n_components` as variable hyperparameters, with `n_components` maybe being optional. You'd then feed the output from UMAP into a clustering method which did not require specification of the number of clusters up front.

This would result in a clustering of the UMAP into a number of classes, with the classes themselves and the number of classes being output from the clustering method. This output could then be fed into a figure-of-merit calculation such as Calinski–Harabasz, which would produce a figure-of-merit which could then be optimised by tuning the hyperparameters.

**Comment by lmcinnes** [comment]
The ``n_neighbors`` parameter in UMAP and the ``n_samples`` parameter in DBSCAN/HDBSCAN mean different things. They are, at least, measured in the same units. There is some reason to believe one could construct a clustering algorithm that would make use of the UMAP ``n_neighbors`` in a similar way to DBSCAN/HDBSCAN. As it stands however, I wouldn't say they should be tied or linked. I would say, however, that they should probably be on roughly the same scale (I would expect them to be the same order of magnitude in general, for instance).

**Comment by bzip2** [comment]
@MattWenham: please do tell.

**Comment by MattWenham** [comment]
Start here for the best description of the algorithm: https://hal.archives-ouvertes.fr/hal-01461451

My implementation use `somoclu` as the SOM engine. It's trapped in a Jupyter notebook at the moment, but I will try to get that shared at some point soon.

**Comment by SamKlier** [comment]
Hi all,
I landed here because I need to cluster high-dimensional data and use UMAP as a preprocessing step before applying HDBSCAN. And unfortunately I can't guess n_neighbors based on the amount of data I have. But of course the results depend heavily on a proper value for n_neighbors (and min_cluster_size).

Therefore, my idea was to "pop" my data into UMAP with a reasonable n_neighbors value, then run HDBSCAN with a similar reasonable min_cluster_size value and validate the result with HDBSCAN's validity_index (which is an approximation of the DBCV metric). Then tune the parameters based on this pipeline. However, the validity_index/DBCV is computationally hardly applicable to high-dimensional data, as it basically raises something like the average density to the power of the data's dimension (if I understand that correctly), which yields numbers too big for float64. So this is where I am at the moment. I need to learn more about DBCV and other metrics that might be appropriate.

Any comments are greatly appreciated!

**Comment by fjatelnicki** [comment]
> Hi all, I landed here because I need to cluster high-dimensional data and use UMAP as a preprocessing step before applying HDBSCAN. And unfortunately I can't guess n_neighbors based on the amount of data I have. But of course the results depend heavily on a proper value for n_neighbors (and min_cluster_size).
> 
> Therefore, my idea was to "pop" my data into UMAP with a reasonable n_neighbors value, then run HDBSCAN with a similar reasonable min_cluster_size value and validate the result with HDBSCAN's validity_index (which is an approximation of the DBCV metric). Then tune the parameters based on this pipeline. However, the validity_index/DBCV is computationally hardly applicable to high-dimensional data, as it basically raises something like the average density to the power of the data's dimension (if I understand that correctly), which yields numbers too big for float64. So this is where I am at the moment. I need to learn more about DBCV and other metrics that might be appropriate.
> 
> Any comments are greatly appreciated!

@SamKlier, did you find any solution for that?

**Comment by bzip2** [comment]
@samklier: HDBSCAN and the validity index are not applicable to the original data in many dimensions (i.e., before UMAP). 

Here's an example. The data that's represented in [this 2D map of cancer](https://depmap.org/portal/celligner/) has >20,000 dimensions. UMAP was used to reduce it to two dimensions. You would run HDBSCAN and evaluate the validity of the resulting clusters in 2D, not in 20,000D.

**Comment by shahadatku** [comment]
I am working on patient clustering using UMAP, where I want to tune the hyperparameters of UMAP using an optimization technique. Is it really possible to do?
